dnf5daemon-server/services/offline/offline.cpp

```diff
@@ -0,0 +1,190 @@
+    auto reply = call.createReply();
+    reply << (offline_transaction_scheduled() == Scheduled::SCHEDULED);
+    return reply;
+}
+
+sdbus::MethodReply Offline::cancel(sdbus::MethodCall & call) {
+    if (!session.check_authorization(dnfdaemon::POLKIT_EXECUTE_RPM_TRANSACTION, call.getSender())) {
+        throw std::runtime_error("Not authorized");
+    }
+    bool success = true;
+    std::string error_msg;
+    switch (offline_transaction_scheduled()) {
+        case Scheduled::SCHEDULED: {
+            std::error_code ec;
+            if (!std::filesystem::remove(libdnf5::offline::MAGIC_SYMLINK, ec) && ec) {
+                success = false;
+                error_msg = ec.message();
+            }
+        } break;
+        case Scheduled::ANOTHER_TOOL:
+            success = false;
+            error_msg = ERR_ANOTHER_TOOL;
+            break;
+        case Scheduled::NOT_SCHEDULED:
+            break;
+    }
+    auto reply = call.createReply();
+    reply << success;
+    reply << error_msg;
+    return reply;
+}
+
+sdbus::MethodReply Offline::clean(sdbus::MethodCall & call) {
+    if (!session.check_authorization(dnfdaemon::POLKIT_EXECUTE_RPM_TRANSACTION, call.getSender())) {
+        throw std::runtime_error("Not authorized");
+    }
+    std::vector<std::string> error_msgs;
+    bool success = true;
+    if (offline_transaction_scheduled() == Scheduled::SCHEDULED) {
+        // remove the magic symlink if it was created by dnf5
+        std::error_code ec;
+        if (!std::filesystem::remove(libdnf5::offline::MAGIC_SYMLINK, ec) && ec) {
+            success = false;
+            error_msgs.push_back(ec.message());
+        }
+    }
+    // clean dnf5 offline transaction files
+    for (const auto & entry : std::filesystem::directory_iterator(get_datadir())) {
+        std::error_code ec;
+        std::filesystem::remove_all(entry.path(), ec);
+        if (ec) {
+            success = false;
+            error_msgs.push_back(ec.message());
+        }
+    }
+    auto reply = call.createReply();
+    reply << success;
+    reply << libdnf5::utils::string::join(error_msgs, ", ");
+    return reply;
+}
+
+sdbus::MethodReply Offline::set_finish_action(sdbus::MethodCall & call) {
+    if (!session.check_authorization(dnfdaemon::POLKIT_EXECUTE_RPM_TRANSACTION, call.getSender())) {
+        throw std::runtime_error("Not authorized");
+    }
+    bool success{false};
+    std::string error_msg{};
+    // try load the offline transaction state
+    const std::filesystem::path state_path{
+        libdnf5::offline::MAGIC_SYMLINK / libdnf5::offline::TRANSACTION_STATE_FILENAME};
+    libdnf5::offline::OfflineTransactionState state{state_path};
+    const auto & read_exception = state.get_read_exception();
+    if (read_exception == nullptr) {
+        // set the poweroff_after item accordingly
+        std::string finish_action;
+        call >> finish_action;
+        state.get_data().set_poweroff_after(finish_action == "poweroff");
```

Comment:
I'm fine to have it like this, even case sensitive, but it might be good to at least test whether the other action is `"reboot"` and claim a warning on the console about "unknown finish action" if it's neither `"poweroff"` nor `"reboot"`. That way one can check for typos and the older library will not crash when a new value is proposed in the future.

Reply:
I can fill the `error_msg` output parameter with such warning. Returning an error reply seems too me.

Reply:
Returning an error, aka failing the call, would make the parameter strict. It's fine by me, it only has certain consequences.